File: src/main/kotlin/com/example/travelator/EmailAddress.java
package com.example.travelator;

public class EmailAddress {
    private final String localPart;
    private final String domain;

    public static EmailAddress parse(String address) {
        var at = address.indexOf('@');
        if (at == -1) {
            throw new IllegalArgumentException("Invalid email address: " + address);
        }
        return new EmailAddress(address.substring(0, at), address.substring(at + 1));
    }

    public EmailAddress(String localPart, String domain) {
        this.localPart = localPart;
        this.domain = domain;
    }

    public String getLocalPart() {
        return localPart;
    }

    public String getDomain() {
        return domain;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return super.equals(obj);
    }

    @Override
    public String toString() {
        return localPart + "@" + domain;
    }
}

File: src/main/kotlin/com/example/travelator/EmailAddress.kt
package com.example.travelator

import java.util.*

class EmailAddress(val localPart: String, val domain: String) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as EmailAddress
        return localPart == that.localPart && domain == that.domain
    }

    override fun hashCode(): Int {
        return Objects.hash(localPart, domain)
    }

    override fun toString(): String {
        return "$localPart@$domain"
    }

    companion object {
        @JvmStatic
        fun parse(value: String): EmailAddress {
            val atIndex = value.lastIndexOf('@')
            require(!(atIndex < 1 || atIndex == value.length - 1)) { "EmailAddress must be two parts separated by @" }
            return EmailAddress(value.substring(0, atIndex), value.substring(atIndex + 1))
        }
    }
}